packages/devtools_app/lib/src/timeline/timeline_model.dart

```diff
@@ -11,27 +15,162 @@ import '../utils.dart';
 
 /// Data model for DevTools Timeline.
-class TimelineData {
-  TimelineData({
+class FrameBasedTimelineData extends TimelineData {
+  FrameBasedTimelineData({
     List<Map<String, dynamic>> traceEvents,
     List<TimelineFrame> frames,
     this.selectedFrame,
+    TimelineEvent selectedEvent,
+    CpuProfileData cpuProfileData,
+    double displayRefreshRate,
+  })  : frames = frames ?? [],
+        _displayRefreshRate = displayRefreshRate,
+        super(
+          traceEvents: traceEvents ?? [],
+          selectedEvent: selectedEvent,
+          cpuProfileData: cpuProfileData,
+        );
+
+  static const displayRefreshRateKey = 'displayRefreshRate';
+
+  static const selectedFrameIdKey = 'selectedFrameId';
+
+  FutureOr<double> get displayRefreshRate async =>
+      _displayRefreshRate ??= await serviceManager.getDisplayRefreshRate();
+
+  double _displayRefreshRate;
+
+  /// All frames currently visible in the timeline.
+  List<TimelineFrame> frames = [];
+
+  TimelineFrame selectedFrame;
+
+  String get selectedFrameId => selectedFrame?.id;
+
+  @override
+  Map<String, dynamic> get json => {
+        selectedFrameIdKey: selectedFrame?.id,
+        displayRefreshRateKey: _displayRefreshRate ?? defaultRefreshRate,
+      }..addAll(super.json);
+
+  @override
+  int get displayDepth =>
+      selectedFrame.uiEventFlow.depth + selectedFrame.gpuEventFlow.depth;
+
+  @override
+  void clear() {
+    super.clear();
+    frames.clear();
+    selectedFrame = null;
+    _displayRefreshRate = null;
+  }
+}
+
+class FullTimelineData extends TimelineData {
+  FullTimelineData({
+    List<Map<String, dynamic>> traceEvents,
+    TimelineEvent selectedEvent,
+    CpuProfileData cpuProfileData,
+    List<TimelineEvent> timelineEvents,
+  })  : timelineEvents = timelineEvents ?? [],
+        super(
+          traceEvents: traceEvents ?? [],
+          selectedEvent: selectedEvent,
+          cpuProfileData: cpuProfileData,
+        );
+
+  static const uiKey = 'UI';
+
+  static const gpuKey = 'GPU';
+
+  static const unknownKey = 'Unknown';
+
+  final List<TimelineEvent> timelineEvents;
+
+  final SplayTreeMap<String, List<TimelineEvent>> eventBuckets =
+      SplayTreeMap(eventBucketComparator);
+
+  TimeRange time = TimeRange();
+
+  @override
+  int get displayDepth {
+    if (_displayDepth != null) return _displayDepth;
+
+    if (eventBuckets.isEmpty) {
+      initializeEventBuckets();
+    }
+
+    int depth = 0;
+    for (List<TimelineEvent> bucket in eventBuckets.values) {
+      int bucketDisplayDepth = 0;
+      for (TimelineEvent event in bucket) {
+        bucketDisplayDepth = max(bucketDisplayDepth, event.displayDepth);
+      }
+      depth += bucketDisplayDepth;
+    }
+    return depth;
+  }
+
+  int _displayDepth;
+
+  void initializeEventBuckets() {
+    for (TimelineEvent event in timelineEvents) {
+      if (event.isAsyncEvent) {
+        eventBuckets.putIfAbsent(event.name, () => [])..add(event);
```

Comment:
simplify this code. Most of these lines are the same just the key is different.
instead have a method to get the 
Compute the key and then call
eventBuckets.putIfAbset(key, () => [])..add(event)

Reply:
done.